### PR TITLE
Google Cloud Storage API [WIP]

### PR DESCRIPTION
**Pull request by hiranya911**
Some code borrowed from #38 

Exposes Google Cloud Storage API via the Java Admin SDK:

* Take a dependency on GCS Node.js client (and the type definitions from DefinitelyTyped)
* Introducing `Storage` type that wraps the GCS storage client
* Expose the GCS buckets via a new `bucket()` method.

go/firebase-admin-gcs

### CODE REVIEW

File: src/storage/storage.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+  }
+
+  public bucket(name?: string): Bucket {
+    let bucketName;
+    if (typeof name !== 'undefined') {
+      bucketName = name;
+    } else {
+      bucketName = this.appInternal.options.storageBucket;
+    }
+    if (typeof bucketName !== 'string' || bucketName === '') {
```
Comment (bojeil-google) [inline src/storage/storage.ts]: Suggestion: you can also use `validator.sNonEmptyString`

File: test/integration/storage.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+var admin = require('../../lib/index');
+var stream = require('stream');
+
+function test(utils) {
+  console.log('\nStorage:');
+  
+  function testDefaultBucket() {
+    const bucket = admin.storage().bucket();
+    return verifyBucket(bucket, 'storage().bucker()')
```
Comment (bojeil-google) [inline test/integration/storage.js]: bucket

File: test/unit/storage/storage.spec.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+'use strict';
+
+import * as _ from 'lodash';
+import {expect} from 'chai';
+
+import * as mocks from '../../resources/mocks';
+import {FirebaseApp} from '../../../src/firebase-app';
+import {Storage} from '../../../src/storage/storage';
+
+describe('Messaging', () => {
```
Comment (bojeil-google) [inline test/unit/storage/storage.spec.ts]: Storage

File: test/unit/storage/storage.spec.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                const bucketAny: any = storage.bucket;
+                bucketAny(invalidName);
+            }).to.throw(expectedError);
+        });
+    });
+  });
+
+  describe('bucket(valid)', () => {
+    it('should return a bucket object', () => {
+      expect(storage.bucket('foo').name).to.equal('foo');
```
Comment (bojeil-google) [inline test/unit/storage/storage.spec.ts]: Should you add another valid test when no bucket name is added?

File: src/storage/storage.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+ */
+class StorageInternals implements FirebaseServiceInternalsInterface {
+  /**
+   * Deletes the service and its associated resources.
+   *
+   * @return {Promise<()>} An empty Promise that will be fulfilled when the service is deleted.
+   */
+  public delete(): Promise<void> {
+    // There are no resources to clean up.
+    return Promise.resolve(undefined);
```
Comment (hiranya911) [inline src/storage/storage.ts]: Done

File: src/storage/storage.ts (hunk 1 above)
Comment (hiranya911) [inline src/storage/storage.ts]: Done

File: test/integration/storage.js (hunk 2 above)
Comment (hiranya911) [inline test/integration/storage.js]: Done

File: test/unit/storage/storage.spec.ts (hunk 3 above)
Comment (hiranya911) [inline test/unit/storage/storage.spec.ts]: Done

File: test/unit/storage/storage.spec.ts (hunk 4 above)
Comment (hiranya911) [inline test/unit/storage/storage.spec.ts]: Done

**Comment by hiranya911** [comment]
Thanks for the feedback. I've made all the suggested changes.

**Pull request merged**